### DISCUSSION TITLE
redirect from server action causes running middleware twice

### DISCUSSION THREAD

**Discussion by hongz1** [vercel/next.js#65900] [Help]
### Summary

I am using nextjs 14.2.3 and found some weird routing flows.

I have a server action that is called from form submission in client component.
When the server action do the job, it redirect to another page.
I could see log that async function was executed correctly and then it redirected.

In middleware, it logs twice and there is POST 303 message between the logs.

Is this intended?
Is this something related to #57257 ?
if so, is #63786 released?

![2024-05-17_144004](https://github.com/vercel/next.js/assets/6565136/456d341a-42e7-4764-8a9d-97419e40cadd)

[UPDATE]
I compared those requests and one is from Next-Router-Prefetch and it has request header "x-action-redirect".
I temporarily added bypassing incoming requests having 'x-action-redirect' request header.

Please someone clarify this whether this routing behavior is intended.

```
if (request.headers.get('x-action-redirect')) return response;
```
 
![2024-05-17_150727](https://github.com/vercel/next.js/assets/6565136/d0a8213a-8687-43cc-b92c-b9b00c98b113)
![2024-05-17_150727](https://github.com/vercel/next.js/assets/6565136/d945a7fa-9e90-4379-8c85-6d6d55815d02)

**Comment by gitlord1234** [comment]
Hi, if this happens to you only in development, the issue might be React Strict Mode which can make pages or endpoints get executed twice. Try adding this into you `next.config.js` in the **root** of your project:

```
import type { NextConfig } from "next";

const nextConfig: NextConfig = {
  reactStrictMode: false,
  /* ... */
};

export default nextConfig;
```

I had redirects in proxy.ts and some server actions and this fixed my problems.
